### DISCUSSION TITLE
837 Loops Issue

### DISCUSSION THREAD

**Discussion by dineshraghuramaiah** [xlate/staedi#221] [Q&A]
Hello,

I am parsing 837 P edi file. 

<--2310A Loop-->
NM1*DN*1*MS.CANDACE LYNN*JONES****XX*1740532381~
<--2310C Loop-->
NM1*77*2*WALKER DALE*****XX*1740532381~
N3*55 BROADWAY STE 2~
N4*BREWER*ME*044121005~

i skipped NM1*82 (2310B Loop) which is optional.

while i am parsing, i am getting 2310B Loop for NM1*77.

is there any way to fix this.

I am expecting 2310C for NM1*77, even though 2310B loop is not coming in edi file.

Thanks,
Dinesh

**Comment by MikeEdgar** [comment]
@dineshraghuramaiah 
Hi Dinesh, how do you have these loops defined in your schema? I assume you're using an implementation definition in the schema.

**Reply by dineshraghuramaiah** [reply]
Yes , i am using schema and implementation.

I am unable to post schema here

**Reply by dineshraghuramaiah** [reply]
<loop maxOccurs="2" minOccurs="0" code="Loop-2310A"
	type="Loop-2310A">

	<sequence>

		<segment maxOccurs="1" minOccurs="1" type="NM1">

		</segment>
		<segment maxOccurs="3" minOccurs="0" type="REF">
		</segment>
		<loop maxOccurs="1" minOccurs="0" code="Loop-2310B"
			type="Loop-2310B">
			<sequence>
				<segment maxOccurs="1" minOccurs="1" type="NM1">
					<sequence>
						<element position="3" title="Hierarchical Level Code (20)">
							<enumeration>
								<value title="Information Source">82</value>
							</enumeration>
						</element>
					</sequence>
				</segment>
				<segment maxOccurs="1" minOccurs="0" type="PRV">
				</segment>
				<segment maxOccurs="4" minOccurs="0" type="REF">
				</segment>
				<loop maxOccurs="1" minOccurs="0" type="Loop-2310C"
					code="Loop-2310C">
					<sequence>
						<segment maxOccurs="1" minOccurs="0" type="NM1">

						</segment>
						<segment maxOccurs="1" minOccurs="0" type="N3">

						</segment>
						<segment maxOccurs="1" minOccurs="0" type="N4">
						</segment>
						<segment maxOccurs="3" minOccurs="0" type="REF">
						</segment>
						<segment maxOccurs="1" minOccurs="0" type="PER">
						</segment>
					</sequence>
				</loop>
			</sequence>
		</loop>
	</sequence>
</loop>

**Reply by dineshraghuramaiah** [reply]
is there any option like if NM101 value is 77 , then loop should be 2310C

**Reply by MikeEdgar** [reply]
@dineshraghuramaiah  A few things about that schema.
1) The `type` for `implementation` loops should be the loop's `code` in the `transaction` schema. Think of the loops in `transaction` as abstract classes and the loops in `implementation` as concrete. So both 2310B and 2310C are of type 2310.
2) In your schema above, you have 2310C *within* 2310B. It should be at the same level. Both loops should specify a `discriminator` attribute of `1` to indicate that NM101 determines which loop it is.

**Reply by dineshraghuramaiah** [reply]
If i have 2130A 2310B and 2310C at same level, for NM1*77* (i.,e., - 2310C) i am getting ref code as 2310A

**Reply by MikeEdgar** [reply]
Do you have the `enumeration` defined for each of the NM1 segments?

**Reply by MikeEdgar** [reply]
You should define each of the 2310 loops like the 1000 loops are defined here: https://github.com/xlate/staedi/blob/master/src/test/resources/x12/005010X222/837_loop1000_only.xml

**Reply by dineshraghuramaiah** [reply]
<loop maxOccurs="2" minOccurs="0" code="Loop-2310A" type="Loop-2310C" discriminator="1">

										<sequence>
											<segment maxOccurs="1" minOccurs="1" type="NM1">
												<sequence>
													<element position="1"
														title="Hierarchical Level Code (20)">
														<enumeration>
															<value title="Information Source">DN</value>
														</enumeration>
													</element>
													<element position="2" />
													<element position="3"
														title="Submitter Last or Organization Name" />
													<element position="4" />
													<element position="5" />
													<element position="8" />
													<element position="9" />
												</sequence>
											</segment>
											<segment maxOccurs="3" minOccurs="0" type="REF">
											</segment>
										</sequence>
									</loop>
									<loop maxOccurs="1" minOccurs="0" code="Loop-2310B"
										type="Loop-2310C" discriminator="1">
										<sequence>
											<segment maxOccurs="1" minOccurs="1" type="NM1">
												<sequence>
													<element position="1"
														title="Hierarchical Level Code (20)">
														<enumeration>
															<value title="Information Source">82</value>
														</enumeration>
													</element>
													<element position="2" />
													<element position="3"
														title="Submitter Last or Organization Name" />
													<element position="4" />
													<element position="5" />
													<element position="8" />
													<element position="9" />
												</sequence>
											</segment>
											<segment maxOccurs="1" minOccurs="0" type="PRV">
											</segment>
											<segment maxOccurs="4" minOccurs="0" type="REF">
											</segment>

										</sequence>
									</loop>
									<loop maxOccurs="1" minOccurs="0" type="Loop-2310C"
										code="Loop-2310C" discriminator="1">
										<sequence>
											<segment maxOccurs="1" minOccurs="0" type="NM1">
												<sequence>
													<element position="1"
														title="Hierarchical Level Code (20)">
														<enumeration>
															<value title="Information Source">77</value>
														</enumeration>
													</element>
													<element position="2" />
													<element position="3"
														title="Submitter Last or Organization Name" />
													<element position="4" />
													<element position="5" />
													<element position="8" />
													<element position="9" />
												</sequence>
											</segment>
											<segment maxOccurs="1" minOccurs="0" type="N3">

											</segment>
											<segment maxOccurs="1" minOccurs="0" type="N4">
											</segment>
											<segment maxOccurs="3" minOccurs="0" type="REF">
											</segment>
											<segment maxOccurs="1" minOccurs="0" type="PER">
											</segment>
										</sequence>
									</loop>

**Reply by dineshraghuramaiah** [reply]
This is out put i am getting. Trying to print as xml nodes.

        <Loop-2310A>
	<Segment-NM1>
		<Element-98>DN</Element-98>
		<Element-1065>1</Element-1065>
		<Element-1035>MS.CANDACE LYNN</Element-1035>
		<Element-1036>JONES</Element-1036>
		<Element-66>XX</Element-66>
		<Element-67>1740532381</Element-67>
	</Segment-NM1>
     </Loop-2310A>
     <Loop-2310A>
	<Segment-NM1>
		<Element-98>77</Element-98>
		<Element-1065>2</Element-1065>
		<Element-1035>WALKER DALE</Element-1035>
		<Element-66>XX</Element-66>
		<Element-67>1740532381</Element-67>
	</Segment-NM1>
	<Segment-N3>
		<null>55 BROADWAY STE 2</null>
	</Segment-N3>
	<Segment-N4>
		<null>BREWER</null>
		<null>ME</null>
		<null>044121005</null>
	</Segment-N4>
     </Loop-2310A>

**Reply by dineshraghuramaiah** [reply]
what i am expecting is like below

     < Loop-2310A>
     <Segment-NM1>
	<Element-98>DN</Element-98>
	<Element-1065>1</Element-1065>
	<Element-1035>MS.CANDACE LYNN</Element-1035>
	<Element-1036>JONES</Element-1036>
	<Element-66>XX</Element-66>
	<Element-67>1740532381</Element-67>
      </Segment-NM1>
           </Loop-2310A>
        <Loop-2310C>
       <Segment-NM1>
	<Element-98>77</Element-98>
	<Element-1065>2</Element-1065>
	<Element-1035>WALKER DALE</Element-1035>
	<Element-66>XX</Element-66>
	<Element-67>1740532381</Element-67>
       </Segment-NM1>
           <Segment-N3>
	<null>55 BROADWAY STE 2</null>
     </Segment-N3>
       <Segment-N4>
	<null>BREWER</null>
	<null>ME</null>
	<null>044121005</null>
           </Segment-N4>
            </Loop-2310C>

**Comment by MikeEdgar** [comment]
@dineshraghuramaiah , I see all three loops are referencing `type="Loop-2310C"`, how is that defined in the `<transaction>` element?

**Reply by dineshraghuramaiah** [reply]
Thanks Mike, i was able to fix it.
